discovery-provider/src/utils/redis_metrics.py

```diff
@@ -57,13 +72,199 @@ def parse_metrics_key(key):
+                timestamp=month,
+                count=unique_monthly_count
+            )
+        session.add(month_unique_record)
+
+        month_total_record = (
+            session.query(MonthlyTotalUsersMetrics)
+            .filter(MonthlyTotalUsersMetrics.timestamp == month)
+            .first()
+        )
+        if month_total_record:
+            month_total_record.count += count
+        else:
+            month_total_record = MonthlyTotalUsersMetrics(
+                timestamp=month,
+                count=count
+            )
+        session.add(month_total_record)
+
+def persist_app_metrics(db, day, month, app_count):
+    with db.scoped_session() as session:
+        for application_name, count in app_count.items():
+            day_record = (
+                session.query(DailyAppNameMetrics)
+                .filter(DailyAppNameMetrics.timestamp == day and \
+                    DailyAppNameMetrics.application_name == application_name)
+                .first()
+            )
+            if day_record:
+                day_record.count += count
+            else:
+                day_record = DailyAppNameMetrics(
+                    timestamp=day,
+                    application_name=application_name,
+                    count=count
+                )
+            session.add(day_record)
+
+            month_record = (
+                session.query(MonthlyAppNameMetrics)
+                .filter(MonthlyAppNameMetrics.timestamp == month and \
+                    MonthlyAppNameMetrics.application_name == application_name)
+                .first()
+            )
+            if month_record:
+                month_record.count += count
+            else:
+                month_record = MonthlyAppNameMetrics(
+                    timestamp=month,
+                    application_name=application_name,
+                    count=count
+                )
+            session.add(month_record)
+
+def merge_metrics(metrics, end_time, metric_type, db):
+    logger.info(f"about to merge {metric_type} metrics: {metrics}")
+    day = end_time.split(':')[0]
+    month = f"{day[:7]}/01"
+
+    daily_key = daily_route_metrics if metric_type == 'route' else daily_app_metrics
+    daily_metrics_str = redis_get_or_restore(REDIS, daily_key)
+    daily_metrics = json.loads(daily_metrics_str) if daily_metrics_str else {}
+
+    monthly_key = monthly_route_metrics if metric_type == 'route' else monthly_app_metrics
+    monthly_metrics_str = redis_get_or_restore(REDIS, monthly_key)
+    monthly_metrics = json.loads(monthly_metrics_str) if monthly_metrics_str else {}
+
+    if day not in daily_metrics:
+        daily_metrics[day] = {}
+    if month not in monthly_metrics:
+        monthly_metrics[month] = {}
+
+    # only relevant for unique users and total api calls
+    unique_daily_count = 0
+    unique_monthly_count = 0
+
+    # only relevant for app metrics
+    app_count = {}
+
+    for new_value, new_count in metrics.items():
```

Comment:
i like how generic you've written these, but maybe a comment explaining new_value here is something like an app name or a ip